[KEYCLOAK-12391] Upgraded select2 and dependent libs

thielef opened the pull request:
- Check out the Jira ticket for the reason for this PR, too
- This upgrades to select2 4.0.10
- Due to that, angular-ui-select2 was not working any more. As this project is obsolete, I used a newer library (also checked vuln. DBs) called ui-select (v0.19.8)
- I also found a problem with redundantly checked in JS dependencies and cleaned this up by deleting themes/src/main/node_modules/
  - Once npm install is run, it puts all under themes/src/main/node_modules/
  - From there, a special npm build of Maven is used to copy it to the target directory to themes/src/main/resources/theme/keycloak/common/resources/node_modules/
  - The latter folder is the one used for all distribution builds, the former one is just temporary and can be removed therefore
- Easier review is possible by splitting your review into all files that contain "node_modules" in their name and all others

stianst commented:
Thanks for your PR. 

How did you update the select2 dependency (did you follow steps in https://github.com/keycloak/keycloak/blob/master/themes/UPDATING-NODE-MODULES.md)? There are a lot more changes in this PR than I would have expected.

thielef commented:
Hi Stian,

I was using "mvn package" and then "mvn process-resources -P npm-profile" to execute the mvn install and copy... target. This copied the newly downloaded npm dependencies (select2 in newer version and ui-select) from themes/src/main/node_modules to themes/src/main/resources/.../node_modules.
Why so many more changes than expected? Because I was unhappy about the duplicated code being checked in. The external dependency is first downloaded to themes/src/main/node_modules and then copied to themes/src/main/resources/.../node_modules. Both folders are fully checked in (and on top are partly listed in .gitignore). Besides the fact that checking in external dependencies seems strange to me, I wanted to clean up at least one step being not required for the normal build process (which only uses the JS libs of the themes/src/main/resources/.../node_modules folder).

BR

Frank

stianst commented:
There are reasons why it is checked in twice. Having them checked in twice means we don't need to run a Maven build to have a working server from the IDE. In the future we plan to improve on this and remove the need to check-in node_modules at all, but there are quite a few factors to consider first.

In either case making changes to a build in the same PR as updating a dependency makes it impossible to review.

thielef commented:
So I shall free the PR from the build changes (i.e. removing the themes/src/main/node_modules/ path) and then you can better review the results?

PS: Maybe you can tell me how to run the server from the IDE without using Maven at all (so that I better understand this) :-)

stianst commented:
Yes, PR without the changes to the build would be great.

To run from an IDE you just need to checkout the source, then you can run KeycloakServer from within the IDE. This is something that will change soon though, as starting early next year we won't commit anything from node_modules, but we can't do that just yet for a range of reasons.

Commit:
Upgraded select2 and dependent libs

thielef commented:
Thanks for the explanation.
Regarding ui-select/select2. The actual removal of my "cleanup" is applied. But the rebase itself seems more problematic due to KEYCLOAK-11472 where many more ui-select2 usages have been introduced... It will probably take until end of next week

thielef commented:
Well, looks like I was faster than expected.

Also done in addition to what I said before:
- During the rebase I also found a bug in kc-component-config.html / RoleLDAPStorageMapperFactory. The ClientID was transferred as single object but had to be put into an array. This causes serialization issues and the initial/stored value could never be loaded.
- I also found out that kc-component-config.html switches the controller itself to the provider controller used for kc-provider-config.html. I applied some slight cleanups (i.e. deleted the related and effectively unused dedicated controller ComponentConfigCtrl) . But I think it makes sense to even merge both html files into one.

thielef commented:
Our internal test found one issue with "Credential reset" for users, checking this

Commit:
Fix

thielef commented:
Issue is fixed

vmuzikar reviewed (changes requested):
@thielef Thank you for your contribution!

I haven't done a full review yet but I ran the tests. It seems like many of the [Admin Console](https://github.com/keycloak/keycloak/blob/master/testsuite/integration-arquillian/HOW-TO-RUN.md#admin-console-ui-tests) tests are broken now – failing with the following exception:
```
NoSuchElementException: no such element: Unable to locate element: {"method":"css selector","selector":"#s2id_reqActions"}
```
Could you please try and fix them? Let me know if I can help.

I'll try to look into this PR more deeply tomorrow.

thielef commented:
Thanks for the test hint. I wasn't aware about them, so I tested all changes manually (such a shame, i know...). I will check these tests now.

thielef commented:
During the tets fix I found another issue, e.g. Cient -> Authorization -> Evaluate -> Roles are directly displayed without searching them

Commit:
Fixed ui-select transformations without special async logic

thielef commented:
At least the issue I found is fixed now in latest commit (d08567c3c160f06ae229755d34a319e3b34573bf)

thielef commented:
And another issue: The type filtering for admins events is broken

Commit:
Fixed broken resourceType admin-event filter

vmuzikar commented:
@thielef Please let me know once this is ready for a test respin. ;)

thielef commented:
@vmuzikar Yes, sure. I am currently working on the AdminConsole tests to make them working:-)

Commit:
Fixed widths for input-group selects

Commit:
Fixed AggregatePolicyManagementTest

Commit:
Fixed AdminEventsTest

thielef commented:
Tests are constantly getting greener, but it takes a while as the xpath support for Selenium is not greatest thing since sliced bread...

Commit:
Fixed ConfigTest

Commit:
Fixed LoginEventsTests

vmuzikar reviewed (changes requested):
I've found more issues.

### Authorization config is broken in IE
```
SCRIPT1002: Syntax error
authz-controller.js (1561,45)
```
```
Error: [$controller:ctrlreg] http://errors.angularjs.org/1.6.10/$controller/ctrlreg?p0=ResourceServerDetailCtrl
   at Anonymous function (http://10.0.2.2:8080/auth/resources/pxldn/admin/keycloak/node_modules/angular/angular.min.js:95:15)
   at link (http://10.0.2.2:8080/auth/resources/pxldn/admin/keycloak/node_modules/angular-route/angular-route.min.js:7:316)
   at Anonymous function (http://10.0.2.2:8080/auth/resources/pxldn/admin/keycloak/node_modules/angular/angular.min.js:17:108)
   at wa (http://10.0.2.2:8080/auth/resources/pxldn/admin/keycloak/node_modules/angular/angular.min.js:86:35)
   at p (http://10.0.2.2:8080/auth/resources/pxldn/admin/keycloak/node_modules/angular/angular.min.js:71:195)
   at f (http://10.0.2.2:8080/auth/resources/pxldn/admin/keycloak/node_modules/angular/angular.min.js:62:477)
   at Anonymous function (http://10.0.2.2:8080/auth/resources/pxldn/admin/keycloak/node_modules/angular/angular.min.js:62:100)
   at Anonymous function (http://10.0.2.2:8080/auth/resources/pxldn/admin/keyc
```

### UI glitches in multi-select fields
* There's some weird extra row under selected items. See e.g. the space below "jboss-logging".
* As a user I'd expect I can click anywhere where's space to add more items (in this case e.g. right next to "jboss-logging"). Currently, to add an item you need to click the blank extra row mentioned above.

![Screenshot 2019-12-16 at 14 43 36](https://user-images.githubusercontent.com/13906855/70911567-786b0080-2012-11ea-8447-3371fd557ae1.png)

This affects many multi-select fields across the Admin Console.

### UI glitches in Client selector
* The hint ("Type at least 1 character to search") is missing some left margin.
![Screenshot 2019-12-16 at 14 51 44](https://user-images.githubusercontent.com/13906855/70912383-1f9c6780-2014-11ea-9f21-38d963c074d5.png)
* The cross for clearing the selection looks a bit odd. It's not well aligned and it uses a different icon than other selectors (big and blue instead black and small).
![Screenshot 2019-12-16 at 14 55 09](https://user-images.githubusercontent.com/13906855/70912430-33e06480-2014-11ea-8e20-2c451057376c.png)

### Licensing
I'm not sure if there's anything we need to do regarding licensing/productization. Perhaps @hmlnarik knows more? :)

Other than that (and apart from failing tests of course) LGTM as far as I can tell. However, it'd be good if someone else with deeper knowledge of our JS codebase could also look into this. @ssilvert?

I'll be away till Jan 2. If there's any progress until then, @mhajas agreed to check it on behalf of me so I wouldn't be blocking the merge.

thielef commented:
Found another issue with "Add scope permission" for clients. Once saved and reloaded and then an scope entry is removed, the remaining items are listed with empty sring...

thielef commented:
One other change to today is that I used "closed-on-select=false" for all multiple selection fields. Besides better bulk support, it has the side effect that for dynamically loaded options (where you have to type in search keys) the option list is preserved between focus changes so that the next time you click into the field you are still getting the full list of old matches. This possibly not relevant list is overwritten when you simple type new keywords. But the user doesn't see the hint that he is  able to do so. Hence, I would like to switch back to the current logic (closed-on-select=true) and make sure that the options are not displayed on when re-entring the field.
Example: Client -> Authorization -> Permissions -> Add scope permission

Commit:
Fixed initialization of multiple ui-select field

Was empty when loaded / one of 2 or more entries was removed after it has been loaded from saved state

Commit:
Fixed ScopePermissionManagementTest

Commit:
Fixed ResourcePermissionManagementTest

Commit:
Fixed ResourceManagementTest

thielef commented:
And found another issue: Initial setup (first clicking in Realm -> Themes onto I14n enabled button ) of locales isn't working (empty list is displayed). It works when setting the locales manually, but the inital load is broken. Going to fix this as well.

Commit:
Fixed I14n initial load of locale options

Commit:
Fixed ClientMappersOIDCTest

Commit:
Fixed ClientClientScopesTest

Commit:
Fixed ClientScopeTest

Also improved ClientClientScopesTest and ClientMappersOIDCTest

Commit:
Fixed RequiredUserActionsTest

thielef commented:
All console tests are green now

thielef commented:
Resolved all issues found by myself so far.
Tomorrow, I will check the so far remaining review comments (i.e. @vmuzikar)

thielef commented:
Regading https://github.com/keycloak/keycloak/pull/6575#issuecomment-566433833 // multiple select fields / close-on-select. I will adapt this as well to simulate the old behaviour and therewith improve the user experience as written in the comment

Commit:
Select option list is reset on selection

thielef commented:
> Regading [#6575 (comment)](https://github.com/keycloak/keycloak/pull/6575#issuecomment-566433833) // multiple select fields / close-on-select. I will adapt this as well to simulate the old behaviour and therewith improve the user experience as written in the comment

Fixed

Commit:
Removed workaround for RequiredUserActionsTest

Commit:
Match highlighting

Commit:
Match highlighting for clientsUiSelect

thielef commented:
Added match highlighting

thielef commented:
Looks like some tests got broken (was pushing too fast). Going to fix them (again :-) )

Commit:
Fixed regressions in tests

thielef commented:
> I've found more issues.
> ### Authorization config is broken in IE
> 
> ```
> SCRIPT1002: Syntax error
> authz-controller.js (1561,45)
> ```
> 
> ```
> Error: [$controller:ctrlreg] http://errors.angularjs.org/1.6.10/$controller/ctrlreg?p0=ResourceServerDetailCtrl
>    at Anonymous function (http://10.0.2.2:8080/auth/resources/pxldn/admin/keycloak/node_modules/angular/angular.min.js:95:15)
>    at link (http://10.0.2.2:8080/auth/resources/pxldn/admin/keycloak/node_modules/angular-route/angular-route.min.js:7:316)
>    at Anonymous function (http://10.0.2.2:8080/auth/resources/pxldn/admin/keycloak/node_modules/angular/angular.min.js:17:108)
>    at wa (http://10.0.2.2:8080/auth/resources/pxldn/admin/keycloak/node_modules/angular/angular.min.js:86:35)
>    at p (http://10.0.2.2:8080/auth/resources/pxldn/admin/keycloak/node_modules/angular/angular.min.js:71:195)
>    at f (http://10.0.2.2:8080/auth/resources/pxldn/admin/keycloak/node_modules/angular/angular.min.js:62:477)
>    at Anonymous function (http://10.0.2.2:8080/auth/resources/pxldn/admin/keycloak/node_modules/angular/angular.min.js:62:100)
>    at Anonymous function (http://10.0.2.2:8080/auth/resources/pxldn/admin/keyc
> ```
Sorry @vmuzikar , but I cannot reproduce this as I have no IE. Also, I cannot find a JS probem in the mentioned line (1561).

thielef commented on themes/src/main/resources-product/theme/rh-sso/admin/theme.properties:
```diff
@@ -1,3 +1,4 @@
 parent=keycloak
 import=common/rh-sso
-styles=node_modules/rcue/dist/css/rcue.min.css node_modules/rcue/dist/css/rcue-additions.min.css node_modules/select2/select2.css css/styles.css lib/angular/treeview/css/angular.treeview.css node_modules/text-security/dist/text-security.css
+styles=node_modules/rcue/dist/css/rcue.min.css node_modules/rcue/dist/css/rcue-additions.min.css node_modules/ui-select/dist/select.css node_modules/select2/dist/css/select2.css css/styles.css lib/angular/treeview/css/angular.treeview.css node_modules/text-security/dist/text-security.css
```
> I've found more issues.
> ### Authorization config is broken in IE
> 
> ```
> SCRIPT1002: Syntax error
> authz-controller.js (1561,45)
> ```
> 
> ```
> Error: [$controller:ctrlreg] http://errors.angularjs.org/1.6.10/$controller/ctrlreg?p0=ResourceServerDetailCtrl
>    at Anonymous function (http://10.0.2.2:8080/auth/resources/pxldn/admin/keycloak/node_modules/angular/angular.min.js:95:15)
>    at link (http://10.0.2.2:8080/auth/resources/pxldn/admin/keycloak/node_modules/angular-route/angular-route.min.js:7:316)
>    at Anonymous function (http://10.0.2.2:8080/auth/resources/pxldn/admin/keycloak/node_modules/angular/angular.min.js:17:108)
>    at wa (http://10.0.2.2:8080/auth/resources/pxldn/admin/keycloak/node_modules/angular/angular.min.js:86:35)
>    at p (http://10.0.2.2:8080/auth/resources/pxldn/admin/keycloak/node_modules/angular/angular.min.js:71:195)
>    at f (http://10.0.2.2:8080/auth/resources/pxldn/admin/keycloak/node_modules/angular/angular.min.js:62:477)
>    at Anonymous function (http://10.0.2.2:8080/auth/resources/pxldn/admin/keycloak/node_modules/angular/angular.min.js:62:100)
>    at Anonymous function (http://10.0.2.2:8080/auth/resources/pxldn/admin/keyc
> ```
Sorry @vmuzikar , but I cannot reproduce this as I have no IE. Also, I cannot find a JS probem in the mentioned line (1561).

thielef replied:
> I'm not sure with this one, but do we really need to specify the select2 styles here again? Isn't it inherited from Keycloak theme?
Hi, @vmuzikar, You might be right. But for whatever reason this has never been cleaned up, not even with the last patternfly changes of @stianst .

thielef commented:
> I've found more issues.
> ### Authorization config is broken in IE
> 
> ```
> SCRIPT1002: Syntax error
> authz-controller.js (1561,45)
> ```
> 
> ```
> Error: [$controller:ctrlreg] http://errors.angularjs.org/1.6.10/$controller/ctrlreg?p0=ResourceServerDetailCtrl
>    at Anonymous function (http://10.0.2.2:8080/auth/resources/pxldn/admin/keycloak/node_modules/angular/angular.min.js:95:15)
>    at link (http://10.0.2.2:8080/auth/resources/pxldn/admin/keycloak/node_modules/angular-route/angular-route.min.js:7:316)
>    at Anonymous function (http://10.0.2.2:8080/auth/resources/pxldn/admin/keycloak/node_modules/angular/angular.min.js:17:108)
>    at wa (http://10.0.2.2:8080/auth/resources/pxldn/admin/keycloak/node_modules/angular/angular.min.js:86:35)
>    at p (http://10.0.2.2:8080/auth/resources/pxldn/admin/keycloak/node_modules/angular/angular.min.js:71:195)
>    at f (http://10.0.2.2:8080/auth/resources/pxldn/admin/keycloak/node_modules/angular/angular.min.js:62:477)
>    at Anonymous function (http://10.0.2.2:8080/auth/resources/pxldn/admin/keycloak/node_modules/angular/angular.min.js:62:100)
>    at Anonymous function (http://10.0.2.2:8080/auth/resources/pxldn/admin/keyc
> ```
Sorry @vmuzikar , but I cannot reproduce this as I have no IE. Also, I cannot find a JS probem in the mentioned line (1561).

thielef commented:
> ### UI glitches in multi-select fields
> 
>     * There's some weird extra row under selected items. See e.g. the space below "jboss-logging".
> 
>     * As a user I'd expect I can click anywhere where's space to add more items (in this case e.g. right next to "jboss-logging"). Currently, to add an item you need to click the blank extra row mentioned above.
> 
> 
> ![Screenshot 2019-12-16 at 14 43 36](https://user-images.githubusercontent.com/13906855/70911567-786b0080-2012-11ea-8447-3371fd557ae1.png)

Hi @vmuzikar , this seems to be a bug in ui-select (looks the same in Plumber, so probably not my integration fault). We could try to fix this by overwriting the in-situ library copy (we "luckily" have):
```
diff --git a/themes/src/main/node_modules/ui-select/dist/select.js b/themes/src/main/node_modules/ui-select/dist/select.js
index 13af545dcb..e77508a8cf 100644
--- a/themes/src/main/node_modules/ui-select/dist/select.js
+++ b/themes/src/main/node_modules/ui-select/dist/select.js
@@ -811,7 +811,7 @@ uis.controller('uiSelectCtrl',
           if (containerWidth === 0) {
             return false;
           }
-          var inputWidth = containerWidth - input.offsetLeft;
+          var inputWidth = containerWidth - input.offsetLeft - 6;
           if (inputWidth < 50) inputWidth = containerWidth;
           ctrl.searchInput.css('width', inputWidth+'px');
           return true;
```
Reasoning: The surrounding div element of the input element (also) has a div element as parent which uses 2 times 3px padding (left and right). This seems to cause the floating issues.
Playing around using CSS (touching both divs) only was not successful (but I am not the most experiences CSS expert...).

PS: The fix of course has to be done in both copies of the file (other one is: themes/src/main/node_modules/ui-select/dist/select.js)

Commit:
Formatting of mulitple select removal link

Commit:
Added missing padding for no-choice hint

thielef commented:
Hi @vmuzikar ,

> ### UI glitches in Client selector
> 
>     * The hint ("Type at least 1 character to search") is missing some left margin.
![Screenshot 2019-12-16 at 14 51 44](https://user-images.githubusercontent.com/13906855/70912383-1f9c6780-2014-11ea-9f21-38d963c074d5.png)
> 
>     * The cross for clearing the selection looks a bit odd. It's not well aligned and it uses a different icon than other selectors (big and blue instead black and small).
![Screenshot 2019-12-16 at 14 55 09](https://user-images.githubusercontent.com/13906855/70912430-33e06480-2014-11ea-8e20-2c451057376c.png)

Fixed both topics with CSS.

thielef commented:
Hi @stianst / @vmuzikar ,

I have replied to and incorporated all feedback and now its time to review again.

vmuzikar commented:
@thielef Thanks for the update!

I've spotted one more UI nitpick. The "arrow down" icon seems a bit off. It seems it's not vertically aligned and it's especially apparent when the cross icon is next to it.
![Screenshot 2020-01-09 at 16 36 28](https://user-images.githubusercontent.com/13906855/72081688-36766700-32ff-11ea-899d-5f7a16fc0855.png)

We still need to address the broken IE. You can download a free VM for testing IE 11 (we don't support other versions) directly [from MS site](https://developer.microsoft.com/en-us/microsoft-edge/tools/vms/).

Regarding the UI glitches in multi-select fields (extra row etc.). I really, really don't like the idea that we'd have to directly modify the dependency sources. Normally I'd say let's create a PR with a fix for the ui-select but unfortunately it seems like a dead unmaintained project (last commit was made almost 2 years ago). This probably leaves us with the only option of embracing this bug. :(
However, I'd like to have a second opinion on this. Perhaps @stianst, could you please review [my comment above](https://github.com/keycloak/keycloak/pull/6575#pullrequestreview-332612275) regarding the glitches in multi-select fields and confirm it's ok to leave them? It's a quite small glitch but it's a glitch we'd introduce with this PR.

One small heads-up. I'm not entirely sure but possibly this PR conflicts with https://github.com/keycloak/keycloak/pull/6432 (which will probably have to be merged sooner).

thielef commented:
> One small heads-up. I'm not entirely sure but possibly this PR conflicts with #6432 (which will probably have to be merged sooner).

OMG, 3400 files have changed... That sounds promising:-)

thielef commented:
> Regarding the UI glitches in multi-select fields (extra row etc.). I really, really don't like the idea that we'd have to directly modify the dependency sources. Normally I'd say let's create a PR with a fix for the ui-select but unfortunately it seems like a dead unmaintained project (last commit was made almost 2 years ago). This probably leaves us with the only option of embracing this bug. :(
> However, I'd like to have a second opinion on this. Perhaps @stianst, could you please review [my comment above](https://github.com/keycloak/keycloak/pull/6575#pullrequestreview-332612275) regarding the glitches in multi-select fields and confirm it's ok to leave them? It's a quite small glitch but it's a glitch we'd introduce with this PR.

As an alternative to adapt their code we could...
- create a fork and publish one artifact which we then us from this fork. An we define this fork to be unmaintained from the beginning on.
- or use a maven step to patch the library before publishing the theme zip (least effort assuming the future theme build doesn't locally store dependencies) -- but the most hacky one.

Commit:
Fixed placement of arrow (same height as X)

thielef commented:
> I've spotted one more UI nitpick. The "arrow down" icon seems a bit off. It seems it's not vertically aligned and it's especially apparent when the cross icon is next to it.
> ![Screenshot 2020-01-09 at 16 36 28](https://user-images.githubusercontent.com/13906855/72081688-36766700-32ff-11ea-899d-5f7a16fc0855.png)

Fixed using CSS.

thielef commented:
> We still need to address the broken IE. You can download a free VM for testing IE 11 (we don't support other versions) directly [from MS site](https://developer.microsoft.com/en-us/microsoft-edge/tools/vms/).

I will give it a try tomorrow / next week...

Commit:
Workaround fixes for Internet Explorer 11

thielef commented:
> One small heads-up. I'm not entirely sure but possibly this PR conflicts with #6432 (which will probably have to be merged sooner).

I jave checked the other PR in more detail. The diff is now smaller (1.4k lines) and seems mergable after some adaptations in pom.xml and repackaging of new npm packages.

vmuzikar commented:
@thielef I discussed with @stianst and @ssilvert the possibilities regarding the ui-select. Unfortunately, we cannot switch to an unmaintained library – it doesn't make much sense to replace one obsolete library with another obsolete library. Do you know if there are any alternative libraries that aren't dead?

I'm sorry about this setback.

abstractj commented:
@stianst @thielef I'm afraid to say that based on the comments above we cannot proceed with the changes based on this PR. I'd suggest us return to the drawing board, close this PR and follow up with the discussions on the dev mailing list. Wdyt?

ssilvert commented:
> @stianst @thielef I'm afraid to say that based on the comments above we cannot proceed with the changes based on this PR. I'd suggest us return to the drawing board, close this PR and follow up with the discussions on the dev mailing list. Wdyt?

It was my understanding from the dev list that we would proceed.   Which is it?

abstractj commented:
@ssilvert based on @vmuzikar comments https://github.com/keycloak/keycloak/pull/6575#issuecomment-575220684 it seems the opposite. I didn't see any threads on the dev mailing list, but if you, @vmuzikar and @stianst have an agreement, that's ok.

thielef commented:
Hi,

I would like to continue the PR by searching an alternative to ui-select.
BTW: Is there a time schedule available for the new admin console?

Mit freundlichen Grüßen / Best regards

Frank Thiele

Open Source Services 2 - Product Group Customer Success Services (INST-CSS/BSV-OS2)
Bosch.IO GmbH | Ullsteinstr. 128 | 12109 Berlin | GERMANY | www.bosch.io<http://www.bosch.io>
Tel. +49 30 726112-0 | Fax +49 30 726112-100 | external.Frank.Thiele@bosch.io<mailto:external.Frank.Thiele@bosch.io>

Sitz: Berlin, Registergericht: Amtsgericht Charlottenburg; HRB 148411 B
Aufsichtsratsvorsitzender: Dr.-Ing. Thorsten Lücke; Geschäftsführung: Dr. Stefan Ferber, Dr. Aleksandar Mitrovic, Yvonne Reckling



Von: Bruno Oliveira da Silva <notifications@github.com>
Gesendet: Dienstag, 21. Januar 2020 20:21
An: keycloak/keycloak <keycloak@noreply.github.com>
Cc: EXTERNAL Thiele Frank (TNG, INST-CSS/BSV-OS2) <external.Frank.Thiele@bosch.io>; Mention <mention@noreply.github.com>
Betreff: Re: [keycloak/keycloak] [KEYCLOAK-12391] Upgraded select2 and dependent libs (#6575)


@stianst<https://github.com/stianst> @thielef<https://github.com/thielef> I'm afraid to say that based on the comments above we cannot proceed with the changes based on this PR. I'd suggest us return to the drawing board, close this PR and follow up with the discussions on the dev mailing list. Wdyt?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/keycloak/keycloak/pull/6575?email_source=notifications&email_token=ANJLC2EMROINY3SBGTLQKBDQ65DJBA5CNFSM4JVLFN32YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJQ5QKQ#issuecomment-576837674>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ANJLC2CWPSUUW6EYNBQDQEDQ65DJBANCNFSM4JVLFN3Q>.

abstractj commented:
@thielef thanks for your response. I will leave it open and up to @ssilvert @stianst and @vmuzikar to decide.

ssilvert commented:
If @vmuzikar says the testing checks out I'm fine with merging it.

I believe that @stianst indicated that we should go ahead.  But maybe wait until after 7.4?

stianst commented:
@thielef although this is not ideal due to the new library not being actively maintained either and also the minor ui glitches @vmuzikar identified, we are okay with merging this PR as is. It's the lesser of two evils I guess. The new admin console will hopefully be ready early 2021, which will be built on PatternFly 4, which includes it's own multi select input field that would replace this one anyways, so hopefully there's no CVEs to this new unmaintained library in the meantime.

vmuzikar commented on themes/src/main/resources/theme/base/admin/resources/partials/user-credentials.html:
```diff
@@ -92,9 +92,12 @@
+                        <ui-select-choices repeat="credType in user.disableableCredentialTypes | filter:$select.search">
+                            <div ng-bind-html="credType | highlight: $select.search"/>
+                        </ui-select-choices>
+                    </ui-select>
```
I'm not sure how to properly review this (without hacking in JS console). It seems to me this field is never displayed since introduction of multi-factor auth. Did we just forgot to remove this? @stianst?

vmuzikar commented on testsuite/integration-arquillian/tests/other/console/src/test/java/org/keycloak/testsuite/console/clients/ClientScopeTest.java:
```diff
@@ -62,7 +62,6 @@ public void before() {
     }
     
     @Test
-    @Ignore //phantomjs sometimes doens't enable "Add Selected >>" button when role is selected
```
This is failing with Chrome. Luckily, there's an easy fix.
https://github.com/keycloak/keycloak/blob/14efa819bd2da1e5ffa55522b95a134f9e168815/testsuite/integration-arquillian/tests/other/console/src/main/java/org/keycloak/testsuite/console/page/roles/RoleCompositeRoles.java#L122
Change it to:
```java
roles.add(getTextFromElement(option).trim());
```

thielef replied:
Will fix it

vmuzikar commented on themes/src/main/resources/theme/base/admin/resources/templates/kc-provider-config.html:
```diff
@@ -27,8 +27,18 @@
+                <ui-select-no-choice>
+                    {{:: 'authz-select-min-character-hint' | translate}}
+                </ui-select-no-choice>
+            </ui-select>
```
This is broken for client mappers (present both in clients and IdPs).

thielef replied:
Will fix it

thielef replied:
It seems to work for RoleLDAPStorageMapper (which I used some time ago to verify this change)

thielef replied:
It looks like it is broken in current keycloak version, too :-)
I fixed it for the RoleLDAPStorageMapper where an array is expected by the backend. But the client mappers (e.g. Audience) do expect a single String. Hence: How to proceed? Looks like on of the mappers must be adapted -> I guess the RoleLdap one as it expects an array where only a normal String is making sense from UI perspective.

vmuzikar replied:
> It looks like it is broken in current keycloak version, too :-)

I didn't investigate it much deeply but I've just checked current master branch and "User Client Role" mapper seem to work fine for me.

thielef replied:
There are 2 type of users for the JS field type. User federation mappers like the RoleLdap one and client mappers like the Audience one. The RoleLdap mapper uses the JS field to express an array field, the Audience one uses the JS field to model a single String field. This must break and it does. Simply try to create a user federation mapper that uses the client id field. This will result in a failure when saving (example here is from console log):
``
09:04:09,835 ERROR [org.keycloak.services.error.KeycloakErrorHandler] (default task-2) Uncaught server error: com.fasterxml.jackson.databind.exc.MismatchedInputException: Cannot deserialize instance of `java.util.ArrayList` out of VALUE_STRING token
 at [Source: (io.undertow.servlet.spec.ServletInputStreamImpl); line: 1, column: 402] (through reference chain: org.keycloak.representations.idm.ComponentRepresentation["config"]->org.keycloak.common.util.MultivaluedHashMap["client.id"])
	at com.fasterxml.jackson.databind.exc.MismatchedInputException.from(MismatchedInputException.java:63)
...
	at org.keycloak.services.filters.KeycloakSessionServletFilter.doFilter(KeycloakSessionServletFilter.java:91)

``

I think it makes sense to fix this by making this field a simple string in all use cases (i.e. adapting the user federation mappers).

vmuzikar reviewed (changes requested):
@thielef Thanks for the update. Unfortunately, after extensive manual testing I've found a few more issues, one of them critical (mappers don't work).

thielef commented:
Something else @stianst  / @vmuzikar : Shall I make it one commit before merging? So for, I fixed all review comments without squashing.

Commit:
Fixed test for Chrome

ssilvert commented:
> Something else @stianst / @vmuzikar : Shall I make it one commit before merging? So for, I fixed all review comments without squashing.

We prefer squashing.

thielef commented:
I will wait until feedback regarding the client mapper issue (@stianst  / @vmuzikar) before rebasing and squashing.

stianst commented:
Since this is starting to get pretty complex and has a few issues that we can't easily resolve I'm starting to wonder if a simpler approach is to patch the select2 version we need, and just include the patched version directly in our repository.

Another option may also be to simply not fix if we can validate that we are not affected by the CVE.

We are planning to rewrite the admin console using React and PatternFly 4 this year, so we just need a stop-gap solution in the interim.

stianst commented:
Reviewed the fixes from select2, there are no code changes, just updates to docs and examples. That means updating the dependency doesn't actually resolve anything. Rather, we should review our usage of select2 to make sure we are not vulnerable. I believe we are not as select2 is used to select something from the server, and the server will barf if you try to select a value that doesn't exist, but that would have to be verified.

stianst reviewed (changes requested):
See comments

stianst commented:
Relevant updates to select2:
https://github.com/select2/select2/issues/4587
https://github.com/select2/select2/issues/5448

stianst commented:
Will review this in more depth next week, but here's some findings:

* Upgrading select2 does not change anything as there is no fix in the underlying library, rather just updates to documentation and examples on how it should be used
* There should be no XSS vulnerability here as we are using Angular (ng-module) to populate the values and it does sanitizing of the values (similarly to what is recommended in updated docs/examples from select2)

With that in mind we will most likely reject this as a non-issue and won't update select2.

stianst commented:
From further verification I can confirm there is no actual fix in Select2 with regards to https://snyk.io/vuln/SNYK-JS-SELECT2-456562. It is simply updating documentation and examples recommending that inputs are sanitised properly.

The dependency is used with the Keycloak admin console, which is built on Angular. Angular automatically performs the required sanitising. As such there is no vulnerability in this regards and we are already doing what the updated documentation and examples from Select2 recommends. As such updating Select2 has no value or affect in this regards.
